ql/src/java/org/apache/hadoop/hive/ql/plan/ConditionalResolverMergeFiles.java

```diff
@@ -419,17 +454,79 @@ private AverageSize getAverageSize(FileSystem inpFs, Path dirPath) {
    */
   private long getMergeSize(FileSystem inpFs, Path dirPath, long avgSize) {
     AverageSize averageSize = getAverageSize(inpFs, dirPath);
-    if (averageSize.getTotalSize() < 0) {
+    return getMergeSize(averageSize.totalSize, averageSize.numFiles, avgSize);
+  }
+
+  private List<FileStatus> getManifestFilePaths(HiveConf conf, Path dirPath) throws IOException {
+    FileSystem manifestFs = dirPath.getFileSystem(conf);
+    List<String> filesKept;
+    List<FileStatus> pathsKept = new ArrayList<>();
+    try (FSDataInputStream inStream = manifestFs.open(new Path(dirPath, Utilities.BLOB_MANIFEST_FILE))) {
+      String paths = IOUtils.toString(inStream, Charset.defaultCharset());
+      filesKept = new ArrayList(Arrays.asList(paths.split(System.lineSeparator())));
+    }
+    // The first string contains the directory information. Not useful.
+    filesKept.remove(0);
+
+    for (String file : filesKept) {
+      pathsKept.add(manifestFs.getFileStatus(new Path(file)));
+    }
+    return pathsKept;
+  }
+
+  private long getMergeSize(List<FileStatus> fileStatuses, long avgSize) {
```

Comment:
can we refactor `AverageSize getAverageSize(FileSystem inpFs, Path dirPath)` and add new signature 
`AverageSize getAverageSize(FileStatus[] fStatus)` instead of code duplication

Reply:
Refactored several methods.